File: benchmarking/benchmark_mt.py
import os
import time
import json
import argparse
import numpy as np
import torch

def run_python_mt_sweeps(out_path):
    print(" ========== Starting Python Multithreaded Benchmarking ========= ")
    step = 10
    NUM_RUNS = 10
    
    with open(out_path, "w") as f:
        N = 10
        while N <= 2000:
            # 1. NumPy MT Benchmark
            A_np = np.ones((N, N), dtype=np.float32)
            B_np = np.ones((N, N), dtype=np.float32) * 2.0
            
            # Warmup
            _ = np.matmul(A_np, B_np)
            
            total_time_np = 0.0
            min_time_np = float('inf')
            
            for r in range(NUM_RUNS):
                start = time.perf_counter()
                _ = np.matmul(A_np, B_np)
                elapsed = time.perf_counter() - start
                total_time_np += elapsed
                if elapsed < min_time_np: min_time_np = elapsed
                
            avg_time_np = total_time_np / NUM_RUNS
            f.write(json.dumps({
                "benchmark": "matmul_mt", "N": N, "kernel": "simd_mt", "lang": "numpy",
                "avg_time": avg_time_np, "min_time": min_time_np
            }) + "\n")
            
            # 2. PyTorch MT Benchmark
            A_pt = torch.ones((N, N), dtype=torch.float32)
            B_pt = torch.ones((N, N), dtype=torch.float32) * 2.0
            
            # Warmup
            _ = torch.matmul(A_pt, B_pt)
            
            total_time_pt = 0.0
            min_time_pt = float('inf')
            
            for r in range(NUM_RUNS):
                start = time.perf_counter()
                _ = torch.matmul(A_pt, B_pt)
                elapsed = time.perf_counter() - start
                total_time_pt += elapsed
                if elapsed < min_time_pt: min_time_pt = elapsed
                
            avg_time_pt = total_time_pt / NUM_RUNS
            f.write(json.dumps({
                "benchmark": "matmul_mt", "N": N, "kernel": "simd_mt", "lang": "pytorch",
                "avg_time": avg_time_pt, "min_time": min_time_pt
            }) + "\n")
            
            f.flush()
            
            if step == 10 and N >= 100:
                step = 100
                print("N stepped up to 100")
            elif step == 100 and N >= 1000:
                step = 1000
                print("N stepped up to 1000")
                
            if N == 2000:
                break
            N += step

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--out", type=str, default="./py-data/mt_benchmark_results.jsonl")
    args = parser.parse_args()
    
    os.makedirs(os.path.dirname(args.out), exist_ok=True)
    run_python_mt_sweeps(args.out)
    print("Done!")

File: benchmarking/benchmark_mt.cpp
#include <cmath>
#include <cstdint>
#include <fstream>
#include <iostream>

#include "./../macrograd/src/macrograd.h"
#include "./../minigrad/src/tensor.h"
#include "profiler.h"

using namespace std;

void run_mt_sweep(ofstream& json_out)
{
    uint step = 10;
    const int NUM_RUNS = 10;
    cout << " ========== Starting Multithreaded Benchmarking ========= " << endl;
    for (size_t N = 10; N <= 2000; N += step)
    {
        double total_time = 0;
        double min_time = 1e9;
        
        // --- 1. C (macrograd) MT ---
        Tensor* c_A = new_tensor(N, N);
        Tensor* c_B = new_tensor(N, N);
        for (size_t i = 0; i < N * N; i++) {
            c_A->data[i] = 1.0f; c_B->data[i] = 2.0f;
        }
        
        Tensor* c_warmup = tensor_matmul_simd_mt(c_A, c_B);
        
        for (int r = 0; r < NUM_RUNS; r++) {
            double start_time = get_wall_time();
            Tensor* c_mt = tensor_matmul_simd_mt(c_A, c_B);
            double time_elapsed = (get_wall_time() - start_time);
            min_time = min(min_time, time_elapsed);
            total_time += time_elapsed;
        }

        double avg_time = total_time / NUM_RUNS;
        json_out << "{\"benchmark\": \"matmul_mt\", \"N\": " << N
                 << ", \"kernel\": \"simd_mt\", \"lang\": \"c\", \"avg_time\": " << avg_time << ", \"min_time\": " << min_time << "}\n";
        g_arena.top = 0; // reset bump allocator

        // --- 2. C++ (minigrad) MT ---
        min_time = 1e9;
        total_time = 0;
        ::Tensor cpp_A(N, N);
        ::Tensor cpp_B(N, N);
        for (size_t i = 0; i < N * N; i++) {
            cpp_A(i/N, i%N) = 1.0f; cpp_B(i/N, i%N) = 2.0f;
        }
        
        ::Tensor cpp_warmup = matmul_simd_mt(cpp_A, cpp_B);
        
        for (int r = 0; r < NUM_RUNS; r++) {
            double start_time = get_wall_time();
            ::Tensor cpp_mt = matmul_simd_mt(cpp_A, cpp_B);
            double time_elapsed = (get_wall_time() - start_time);
            min_time = min(min_time, time_elapsed);
            total_time += time_elapsed;
        }

        avg_time = total_time / NUM_RUNS;
        json_out << "{\"benchmark\": \"matmul_mt\", \"N\": " << N
                 << ", \"kernel\": \"simd_mt\", \"lang\": \"cpp\", \"avg_time\": " << avg_time << ", \"min_time\": " << min_time << "}\n";

        if (step == 10 && N >= 100) step = 100;
        if (step == 100 && N >= 1000) step = 1000;
    }
}

int main()
{
    ofstream json_out("./cpp-data/mt_benchmark_results.jsonl");
    run_mt_sweep(json_out);
    json_out.close();
    return 0;
}
